apps/ejabberd/src/ejabberd_auth.erl

```diff
@@ -218,18 +218,12 @@ do_check_password_loop([AuthModule | AuthModules], Args) ->
     end.
 
 -spec check_digest(binary(), fun(), binary(), binary()) -> boolean().
-check_digest(Digest, DigestGen, Password, Passwd) ->
-    DigRes = if
-                 Digest /= <<>> ->
-                     Digest == DigestGen(Passwd);
-                 true ->
-                     false
-             end,
-    if DigRes ->
-           true;
-       true ->
-           (Passwd == Password) and (Password /= <<>>)
-    end.
+check_digest(<<>>, _, <<>>, _) ->
+    false; %%empty digest and password
+check_digest(<<>>, _, Password, Passwd) ->
```

Comment:
> what do you think about the following clause: check_digest(<<>>, _, Password, Password) -> true; ?

Yeah, that will be better :)

> Moreover, there is something wrong with this function - It says check_digest, but it also checks passwords ... Maybe it is a good moment to clean this mess.

I know I know... Some day will finally (hopefully) get rid of all the parent's mess. I'll refactor this function in this PR.